### PR TITLE
feat: added vote collector, refactored broadcast

### PR DESCRIPTION
**Pull request by randyahx**
### Description

added a vote collector that queries the blockchain for votes using eventhash and stores them for 2/3 consensus.  
refactored broadcast to collate votes using eventhash instead of challengeid, broadcast votes periodically until expired height  
refactored broadcaster, collator, submitter to use for loop instead of goroutine to solve performance issues  
refactored workflow for collector, broadcaster, collator, submitter  
added db wiper to periodically clear db   

### Rationale

services could not sync their votes to submit a msgattest  

### Example

none  

### Changes

Notable changes: 
vote collector
vote broadcast  
models, dao
data providers

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -3,6 +3,10 @@ package app
 import (
 	"fmt"
 
+	"github.com/bnb-chain/greenfield-challenger/logging"
```
Comment (forcodedancing) [inline app/app.go:6]: sort imports please

File: app/app.go [hunk 2]
```diff
@@ -12,24 +16,38 @@ import (
...
-		panic(fmt.Sprintf("open db error, err=%s", err.Error()))
+		panic(fmt.Sprintf("open db error, err=%+v", err.Error()))
+	}
+
+	dbConfig, err := db.DB()
+	if err != nil {
+		panic(err)
 	}
+	dbConfig.SetMaxIdleConns(20)
+	dbConfig.SetMaxOpenConns(40)
```
Comment (forcodedancing) [inline app/app.go:44]: Can we add these parameters into config?

File: app/app.go [hunk 3]
```diff
@@ -12,24 +16,38 @@ import (
...
+
+	err = ResetDB(db, &model.Block{}, &model.Event{}, &model.Vote{})
```
Comment (forcodedancing) [inline app/app.go:46]: Do not `ResetDB` if it is not debugging.

File: app/app.go [hunk 4]
```diff
@@ -47,27 +65,44 @@ func NewApp(cfg *config.Config) *App {
...
+	go a.executor.GetHeightLoop()
 	go a.eventMonitor.ListenEventLoop()
 	go a.hashVerifier.VerifyHashLoop()
-	go a.voteProcessor.SignVoteLoop()
-	go a.voteProcessor.CollectVotesLoop()
-	a.txSubmitter.SubmitTransactionLoop()
+	go a.voteCollector.CollectVotesLoop()
+	go a.voteBroadcaster.BroadcastVotesLoop()
+	go a.voteCollator.CollateVotesLoop()
+	go a.txSubmitter.SubmitTransactionLoop()
```
Comment (forcodedancing) [inline app/app.go:96]: The last loop should not run in a goroutine, otherwise the program will finish.

File: app/app_test.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package app
+
+import (
```
Comment (forcodedancing) [inline app/app_test.go:3]: Move it to another place, or remove it if it is only for debug.

File: common/const.go [hunk 6]
```diff
@@ -7,9 +7,10 @@ import (
...
-	RtyAttem      = retry.Attempts(RtyAttemNum)
-	RtyDelay      = retry.Delay(time.Millisecond * 400)
-	RtyErr        = retry.LastErrorOnly(true)
-	RetryInterval = 2 * time.Second
+	RtyAttemNum        = uint(5)
+	RtyAttem           = retry.Attempts(RtyAttemNum)
+	RtyDelay           = retry.Delay(time.Millisecond * 400)
+	RtyErr             = retry.LastErrorOnly(true)
+	RetryInterval      = 1 * time.Second
+	CacheClearInterval = 1000
```
Comment (forcodedancing) [inline common/const.go:15]: The name is not very clear - it seems to be the times of iteration.

File: config/config.json [hunk 7]
```diff
@@ -9,7 +9,7 @@
     "grpc_addrs": [
       "gnfd-dev-grpc.qa.bnbchain.world:9090"
     ],
-    "private_key": "cf1366138ffd14328db680c57db7d56471017e57a5ae203ca0f3fa84cb0b8ebb",
+    "private_key": "a5ae825a4d0f6e7ddea8823b76fba0357b9c31d6a9965bc9df00300bd3445bad",
```
Comment (forcodedancing) [inline config/config.json:12]: Please refer to relayer's pr for reading private keys.

File: db/dao/event_dao_test.go [hunk 8]
```diff
@@ -96,30 +96,30 @@ func (s *eventSuite) TestEventDao_GetEarliestEventByStatus() {
...
-	result, err = s.dao.GetEarliestEventsByStatus(model.Unprocessed, 1)
-	s.Require().NoError(err, "failed to query")
-	s.Require().True(len(result) == 1)
-	s.Require().True(result[0].ChallengeId == 1)
-
-	result, err = s.dao.GetEarliestEventsByStatus(model.Verified, 10)
-	s.Require().NoError(err, "failed to query")
-	s.Require().True(len(result) == 1)
-	s.Require().True(result[0].ChallengeId == 100)
+	//
```
Comment (forcodedancing) [inline db/dao/event_dao_test.go:102]: Can we fix these tests instead of commenting them out.

File: db/model/event.go [hunk 9]
```diff
@@ -14,8 +14,9 @@ type Event struct {
 	ChallengerAddress string       `gorm:"NOT NULL"`
 	Height            uint64       `gorm:"NOT NULL;"`
 	Status            EventStatus  `gorm:"NOT NULL;index:idx_status"`
-	VerifyResult      VerifyResult `gorm:"NOT NULL;"`
+	VerifyResult      VerifyResult `gorm:"NOT NULL;index:idx_verify_result"`
 	CreatedTime       int64        `gorm:"NOT NULL"`
+	ExpiredHeight     uint64       `gorm:"NOT NULL;index:expired_height"`
```
Comment (forcodedancing) [inline db/model/event.go:19]: Rename the index to `idx_expired_height` to align with others.

Comment (forcodedancing) [inline db/model/event.go:19]: Status `EnoughVotesCollected` is not used now.

File: executor/executor.go [hunk 10]
```diff
@@ -10,6 +10,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/bnb-chain/greenfield-challenger/common"
```
Comment (forcodedancing) [inline executor/executor.go:13]: sort the improts

Comment (forcodedancing) [inline executor/executor.go:13]: sort the imports

Comment (forcodedancing) [inline executor/executor.go:13]: sort the imports

File: monitor/monitor.go [hunk 11]
```diff
@@ -144,17 +151,34 @@ func (m *Monitor) monitorChallengeEvents(block *tmtypes.Block, blockResults *cty
 		BlockTime:   block.Time.Unix(),
 		CreatedTime: time.Now().Unix(),
 	}
-	return m.daoManager.SaveBlockAndEvents(b, EntitiesToDtos(uint64(block.Height), events))
+	err = m.daoManager.SaveBlockAndEvents(b, EntitiesToDtos(uint64(block.Height), events))
+	for _, event := range events {
```
Comment (forcodedancing) [inline monitor/monitor.go:155]: Change the log level to `Debug` for it is for debug purpose.

File: submitter/tx_submitter.go [hunk 12]
```diff
@@ -34,119 +36,108 @@ func NewTxSubmitter(cfg *config.Config, executor *executor.Executor, daoManager
 }
 
 func (s *TxSubmitter) SubmitTransactionLoop() {
+	s.cachedEventHash = make(map[uint64][]byte, math.MaxInt)
+	submitAttempt := 0
```
Comment (forcodedancing) [inline submitter/tx_submitter.go:40]: The name `submitAttempt` is a bit misleading here. It is used to clear/purge the cache after some loop times. It is not about retry or submit attempt.

Comment (forcodedancing) [inline submitter/tx_submitter.go:40]: As we discussed, the logic will change a bit after the in-turn challenger on chain.

File: verifier/hash_verifier.go [hunk 13]
```diff
@@ -4,11 +4,15 @@ import (
...
 	"context"
 	"encoding/hex"
+	"fmt"
 	"io"
+	"math"
 	"strings"
 	"sync"
 	"time"
 
+	"github.com/panjf2000/ants/v2"
```
Comment (forcodedancing) [inline verifier/hash_verifier.go:14]: There are some `TODO: revert` in this file. Could you check and fix them?

File: vote/data_provider.go [hunk 14]
```diff
@@ -37,44 +30,18 @@ func NewDataHandler(daoManager *dao.DaoManager, executor *executor.Executor) *Da
...
-	bs = append(bs, spOperatorBz...)
-	bs = append(bs, challengerBz...)
-	hash := sdk.Keccak256Hash(bs)
-	return hash
-}
-
-func (h *DataHandler) FetchEventsForSelfVote() ([]*model.Event, error) {
-	events, err := h.daoManager.GetEarliestEventsByStatusAndAfter(model.Verified, batchSize, h.lastIdForSelfVote)
+func (h *DataHandler) FetchEventsForSelfVote(currentHeight uint64) ([]*model.Event, error) {
+	// TODO: Revert this
```
Comment (forcodedancing) [inline vote/data_provider.go:34]: Please check this.

File: vote/vote_broadcaster.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
+package vote
+
+import (
+	"fmt"
+	"github.com/bnb-chain/greenfield-challenger/common"
```
Comment (forcodedancing) [inline vote/vote_broadcaster.go:5]: sort the imports

File: vote/vote_broadcaster.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+		signer:          signer,
+		executor:        executor,
+		DataProvider:    kind,
+		cachedLocalVote: nil,
+		blsPublicKey:    getBlsPubKeyFromPrivKeyStr(cfg.VotePoolConfig.BlsPrivateKey),
+	}
+}
+
+func (p *VoteBroadcaster) BroadcastVotesLoop() {
+	p.cachedLocalVote = make(map[uint64]*votepool.Vote, math.MaxInt)
```
Comment (forcodedancing) [inline vote/vote_broadcaster.go:43]: `math.MaxInt` is too big.

File: vote/vote_broadcaster.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+		}
+
+		time.Sleep(RetryInterval)
+	}
+}
+
+func (p *VoteBroadcaster) broadcastForSingleEvent(localVote *votepool.Vote, event *model.Event) error {
+	err := p.preCheck(event)
+	if err != nil {
+		if strings.Contains(err.Error(), "event expired") {
```
Comment (forcodedancing) [inline vote/vote_broadcaster.go:95]: Can you define an error to replace this string, to make it clearer? `"event expired"`

File: vote/vote_collector.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+	"sync"
+	"time"
+
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/bnb-chain/greenfield-challenger/config"
+	"github.com/bnb-chain/greenfield-challenger/db/dao"
+	"github.com/bnb-chain/greenfield-challenger/executor"
+	"github.com/bnb-chain/greenfield-challenger/logging"
+	"github.com/tendermint/tendermint/votepool"
```
Comment (forcodedancing) [inline vote/vote_collector.go:15]: sort imports please

File: vote/vote_collector.go [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+		if err := verifySignature(v, v.EventHash); err != nil {
+			logging.Logger.Errorf("verify vote's signature failed,  err=%+v", err)
+			continue
+		}
+
+		err = p.daoManager.SaveVote(EntityToDto(v, uint64(0)))
+		if err != nil {
+			return err
+		}
+		logging.Logger.Infof("vote saved: %s", string(v.Signature))
```
Comment (forcodedancing) [inline vote/vote_collector.go:92]: `string(v.Signature)` is not readable. Can we refine this log?

File: db/model/event.go (hunk 9 above)
Comment (randyahx) [reply db/model/event.go:19]: fixed

Comment (randyahx) [reply db/model/event.go:19]: fixed

File: app/app.go (hunk 2 above)
Comment (randyahx) [reply app/app.go:44]: fixed

File: app/app.go (hunk 4 above)
Comment (randyahx) [reply app/app.go:96]: fixed

File: app/app_test.go (hunk 5 above)
Comment (randyahx) [reply app/app_test.go:3]: fixed

File: common/const.go (hunk 6 above)
Comment (randyahx) [reply common/const.go:15]: fixed

File: db/dao/event_dao_test.go (hunk 8 above)
Comment (randyahx) [reply db/dao/event_dao_test.go:102]: fixed

File: submitter/tx_submitter.go [hunk 20]
```diff
@@ -34,119 +36,108 @@ func NewTxSubmitter(cfg *config.Config, executor *executor.Executor, daoManager
 }
 
 func (s *TxSubmitter) SubmitTransactionLoop() {
+	s.cachedEventHash = make(map[uint64][]byte, math.MaxInt)
```
[the remark this answers is not in the document]
Comment (randyahx) [reply submitter/tx_submitter.go:39]: fixed

File: submitter/tx_submitter.go (hunk 12 above)
Comment (randyahx) [reply submitter/tx_submitter.go:40]: fixed

File: vote/vote_broadcaster.go (hunk 16 above)
Comment (randyahx) [reply vote/vote_broadcaster.go:43]: fixed

File: vote/vote_collector.go (hunk 19 above)
Comment (randyahx) [reply vote/vote_collector.go:92]: fixed

File: app/app.go (hunk 1 above)
Comment (randyahx) [reply app/app.go:6]: fixed

File: executor/executor.go (hunk 10 above)
Comment (randyahx) [reply executor/executor.go:13]: fixed

File: vote/vote_collector.go (hunk 18 above)
Comment (randyahx) [reply vote/vote_collector.go:15]: fixed

File: app/app.go [hunk 21]
```diff
@@ -45,29 +75,69 @@ func NewApp(cfg *config.Config) *App {
...
-		voteProcessor: voteProcessor,
-		txSubmitter:   txSubmitter,
+		executor:        executor,
+		eventMonitor:    monitor,
+		hashVerifier:    hashVerifier,
+		voteCollector:   voteCollector,
+		voteBroadcaster: voteBroadcaster,
+		voteCollator:    voteCollator,
+		txSubmitter:     txSubmitter,
+		dbWiper:         dbWiper,
```
Comment (forcodedancing) [inline app/app.go:97]: Do we need wiper? Sometimes we need to review or query historical data. We may not need it now.

**Pull request merged**